digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/TrafficLightService.scala

```diff
@@ -34,4 +37,87 @@ class TrafficLightService(val roadLinkService: RoadLinkService) extends PointAss
       OracleTrafficLightDao.create(TrafficLightToBePersisted(asset.linkId, asset.lon, asset.lat, mValue, municipality, username), username)
     }
   }
+
+  override def getByBoundingBox(user: User, bounds: BoundingRectangle): Seq[PersistedAsset] = {
+    case class AssetBeforeUpdate(asset: PersistedAsset, persistedFloating: Boolean, floatingReason: Option[FloatingReason])
+    val (roadLinks, changeInfo) = roadLinkService.getRoadLinksAndChangesFromVVH(bounds)
+
+    withDynTransaction {
+      val boundingBoxFilter = OracleDatabase.boundingBoxFilter(bounds, "a.geometry")
+      val filter = s"where a.asset_type_id = $typeId and $boundingBoxFilter"
+      val persistedAssets: Seq[PersistedAsset] = fetchPointAssets(withFilter(filter), roadLinks)
+
+      val assetsBeforeUpdate: Seq[AssetBeforeUpdate] = persistedAssets.filter { persistedAsset =>
+        user.isAuthorizedToRead(persistedAsset.municipalityCode)
+      }.map { (persistedAsset: PersistedAsset) =>
+        val (floating, assetFloatingReason) = super.isFloating(persistedAsset, roadLinks.find(_.linkId == persistedAsset.linkId))
+
+        if (floating || persistedAsset.floating) {
+          PointAssetFiller.correctedPersistedAsset(persistedAsset, roadLinks, changeInfo) match {
+            case Some(trafficLight) =>
+              OracleTrafficLightDao.update(trafficLight.assetId, TrafficLightToBePersisted(trafficLight.linkId,
+                trafficLight.lon, trafficLight.lat, trafficLight.mValue, persistedAsset.municipalityCode, "vvh_generated"))
+
+              AssetBeforeUpdate(new PersistedAsset(trafficLight.assetId, trafficLight.linkId, trafficLight.lon, trafficLight.lat,
+                trafficLight.mValue, trafficLight.floating, persistedAsset.municipalityCode, persistedAsset.createdBy,
+                persistedAsset.createdAt, persistedAsset.modifiedBy, persistedAsset.modifiedAt), trafficLight.floating, Some(FloatingReason.Unknown))
+
+            case None =>
+              if (floating && !persistedAsset.floating) {
+                val logger = LoggerFactory.getLogger(getClass)
+                val floatingReasonMessage = floatingReason(persistedAsset, roadLinks.find(_.linkId == persistedAsset.linkId))
+                logger.info("Floating asset %d, reason: %s".format(persistedAsset.id, floatingReasonMessage))
+              }
+              AssetBeforeUpdate(setFloating(persistedAsset, floating), persistedAsset.floating, assetFloatingReason)
+          }
+        }
+        else
+          AssetBeforeUpdate(setFloating(persistedAsset, floating), persistedAsset.floating, assetFloatingReason)
+      }
+      assetsBeforeUpdate.foreach { asset =>
+        if (asset.asset.floating != asset.persistedFloating) {
+          updateFloating(asset.asset.id, asset.asset.floating, asset.floatingReason)
+        }
+      }
+      assetsBeforeUpdate.map(_.asset)
+    }
+  }
+
+  override def getByMunicipality(municipalityCode: Int): Seq[PersistedAsset] = {
```

Comment:
It seems like there is a lot of duplicated code we can have a method on the base class that has a function as a parameter, and then reuse it.